pkg/macaroon/serialize.go

```diff
@@ -0,0 +1,205 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package macaroon
+
+import (
+	"encoding/binary"
+	"errors"
+)
+
+type fieldType int
+
+const (
+	fieldEOS            fieldType = 0
+	fieldLocation       fieldType = 1
+	fieldIdentifier     fieldType = 2
+	fieldVerificationID fieldType = 4
+	fieldSignature      fieldType = 6
+)
+
+type packet struct {
+	fieldType fieldType
+	data      []byte
+}
+
+// Serialize converts macaroon to binary format
+func (m *Macaroon) Serialize() (data []byte) {
+	// Start data from version int
+	data = append(data, 2)
+
+	// Serilize Identity
+	data = serializePacket(data, packet{
+		fieldType: fieldIdentifier,
+		data:      m.head,
+	})
+	data = append(data, 0)
+
+	// Serialize caveats
+	for _, cav := range m.caveats {
+		data = serializePacket(data, packet{
+			fieldType: fieldIdentifier,
+			data:      cav,
+		})
+		data = append(data, 0)
+	}
+
+	data = append(data, 0)
+
+	// Serialize tail
+	data = serializePacket(data, packet{
+		fieldType: fieldSignature,
+		data:      m.tail,
+	})
+
+	return data
+}
+
+// serializePacket converts packet to binary
+func serializePacket(data []byte, p packet) []byte {
+	data = appendVarint(data, int(p.fieldType))
+	data = appendVarint(data, len(p.data))
+	data = append(data, p.data...)
+
+	return data
+}
+
+func appendVarint(data []byte, x int) []byte {
+	var buf [binary.MaxVarintLen32]byte
```

Comment:
No need that amount of allocation, considering that `uint64` is used and per https://golang.org/ref/spec#Size_and_alignment_guarantees with 8 is enough.